text/0000-custom-dst.md

```diff
@@ -0,0 +1,479 @@
+- Feature Name: `custom_dst`
+- Start Date: 2018-11-06
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Allow Rust code to define dynamically sized types with custom thick (and thin)
+pointers, and define slice functions in terms of these, instead of transmute.
+Also, convert the `CStr` type to use this functionality,
+and make it a thin pointer; this will allow use with FFI.
+
+# Motivation
+[motivation]: #motivation
+
+As of right now, the lack of custom DSTs in Rust means that we can't communicate
+with C in important ways - we lack the ability to define a `CStr` in the
+language such that `&CStr` is compatible with `char const *`,
+and we lack the ability to communicate nicely with C code that uses
+[Flexible Array Members](https://en.wikipedia.org/wiki/Flexible_array_member).
+This RFC attempts to fix this,
+as well as introduce more correctness to existing practices.
+
+Apart from FFI, it also has usecases for indexing and slicing 2-d arrays.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+There's a new language trait in the standard library, similar to `Sized`,
+under `std::marker`:
+
+```rust
+unsafe trait DynamicallySized {
+    /*
+        note: these are all required due to trait implementations for
+          - Unpin - all pointer types
+          - Copy + Send + Sync - &T, &mut T
+          - Eq + Ord - *const T, *mut T
+    */
+    type Metadata: 'static + Copy + Send + Sync + Eq + Ord + Unpin;
+
+    fn size_of_val(&self) -> usize;
+    fn align_of_val(meta: Self::Metadata) -> usize;
+}
+```
+
+with an automatic implementation for all `Sized` types:
+
+```rust
+// note: this is _only_ for explanation
+// this should happen in the compiler
+unsafe impl<T> DynamicallySized for T {
+    type Metadata = ();
+
+    fn size_of_val(&self) -> usize { size_of::<T>() }
+    fn align_of_val((): ()) -> usize { align_of::<T>() }
+}
+```
+
+If you have a type which you would like to be unsized,
+you can implement this trait for your type!
+
+```rust
+#[repr(C)]
+struct CStr([c_char; 0]);
+
+unsafe impl DynamicallySized for CStr {
```

Comment:
One weird quirk: `&dyn DynamicallySized` is nonsense because the metadata has unknown size, so we can't have a reference containing the metadata. Clearly the bound here to express that isn't `: Sized`. What is it, exactly? Is the fact that this bound is inexpressible just a quirk that we paper over via compiler builtins? cc @eddyb who may have thought about this before.